tutorials/tensordictmodule.ipynb

```diff
@@ -0,0 +1,1060 @@
+      "File \u001b[0;32m~/anaconda3/envs/torch_rl/lib/python3.9/site-packages/torch/nn/modules/linear.py:114\u001b[0m, in \u001b[0;36mLinear.forward\u001b[0;34m(self, input)\u001b[0m\n\u001b[1;32m    113\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mforward\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;28minput\u001b[39m: Tensor) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Tensor:\n\u001b[0;32m--> 114\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mF\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlinear\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mweight\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbias\u001b[49m\u001b[43m)\u001b[49m\n",
+      "\u001b[0;31mRuntimeError\u001b[0m: batched == nullptr INTERNAL ASSERT FAILED at \"/private/var/folders/fn/c72nxv0x0xj4bzdgq3b0r86r0000gn/T/pip-req-build-ulv3uwfj/functorch/csrc/Interpreter.cpp\":95, please report a bug to PyTorch. "
+     ]
+    }
+   ],
+   "source": [
+    "num_models = 10\n",
+    "\n",
+    "tensordict = TensorDict({\"a\": torch.randn(5, 3)}, batch_size=[5])\n",
+    "\n",
+    "splitlinear_models = [\n",
+    "    TensorDictModule(\n",
+    "        nn.Linear(3, 10), in_keys=[\"a\"], out_keys=[\"output\"]\n",
+    "    )\n",
+    "    for _ in range(num_models)\n",
+    "]\n",
+    "\n",
+    "\n",
+    "def transpose_stack(tuple_of_tuple_of_tensors):\n",
+    "    tuple_of_tuple_of_tensors = tuple(zip(*tuple_of_tuple_of_tensors))\n",
+    "    results = tuple(\n",
+    "        torch.stack(shards) for shards in tuple_of_tuple_of_tensors\n",
+    "    )\n",
+    "    return results\n",
+    "\n",
+    "\n",
+    "### Let's extract the functional version of the common model\n",
+    "func = splitlinear_models[0].make_functional_with_buffers()[0]\n",
+    "\n",
+    "### We also extract parameters and buffers for every block and stack them together\n",
+    "params, buffers = zip(\n",
+    "    *[\n",
+    "        splitlinear.make_functional_with_buffers()[1]\n",
+    "        for splitlinear in splitlinear_models\n",
+    "    ]\n",
+    ")\n",
+    "params = transpose_stack(params)\n",
+    "buffers = transpose_stack(buffers)\n",
+    "\n",
+    "### Finally we can apply the function to the stack of params and buffer using vmap\n",
+    "func(tensordict, params=params, buffers=buffers, vmap=True).shape"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "31be6c45-10fb-4fd1-a52f-92214b76c00a",
+   "metadata": {},
+   "source": [
+    "## Do's and don't with `TensorDictModule`\n",
```

Comment:
Let's do a proper list here. Add examples to be clear.

e.g.: 
**don't** use a different name for the output of a TensorDictModule as the output tensordict is just the input modified in-place:
```python
tensordict = module(tensordict)  # ok!
tensordict_out = module(tensordict)  # don't!
```

**don't** use `nn.Sequence` but TensorDictSequence. _show that make_functional_with_wrapper does not work, and that you can't access the in_keys_

Reply:
etc. 
Try to find a couple more :)